### DISCUSSION TITLE
Ionic fingerprint dialog is crashing in android

### DISCUSSION THREAD

**Discussion by granith** [NiklasMerz/cordova-plugin-fingerprint-aio#111] [Q&A]
I have used this https://ionicframework.com/docs/native/fingerprint-aio/, for implementing touchID and FaceId for my app. In IOS it is working perfectly but when I emulate my app in Android device fingerprint dialog is not showing and app is just crashing. I have used same code as it is described in link above.

```
    this.faio.isAvailable().then(result =>{
      this.faio.show({
        clientId: 'MyBox app authenication',
        clientSecret: 'password'
      })
      .then((result: any) =>{
        console.log("Success";)
          })
          .catch((error: any) => {
            this.message = true;
            this.navCtrl.setRoot(LoginPage);
          });
        }})
      .catch((error: any) => {
        this.errorfingerprint = true;
      });
    }).catch(err => {
        // alert("No TouchId or FaceID in this phone");
    });
```
Any idea what it can be? I followed every instruction but no success.

**Comment by NiklasMerz** [comment]
Did you check adb, Cordova logs, and logcat? Maybe your phone is not setup correcty.

**Comment by granith** [comment]
Yes, I have checked them, everything seems to be good, I have tried with 3 different phones but same error. Dialogue box only appears in next screen after I log in.

**Comment by NiklasMerz** [comment]
This sounds like an error in your app. Check if you are doing everything in the right order. For questions please use Gitter in favor of issues on this repo.

**Comment by akildemir** [comment]
I am having the same issue. Everything seems to be correct but my phone(android) just doesn't show the fingerprint authentication dialog. My code is : `fingerprintOptions : FingerprintOptions;

  constructor(public navCtrl: NavController, public navParams: NavParams, private platforn : Platform, private fingerprint : FingerprintAIO) {

    this.fingerprintOptions = {

      clientId : 'demo',
      clientSecret : 'password',
      disableBackup : true


    }

async verify(){

    try{

      await this.platforn.ready();
      const ava =await this.fingerprint.isAvailable();
      if(ava === "OK")
      {
        const res = await this.fingerprint.show(this.fingerprintOptions);
        console.log(res);
      }

    }catch(e){

        console.log(e);

    }


`

**Comment by granith** [comment]
I solved this problem by removing this plugin https://ionicframework.com/docs/native/keychain-touch-id/. Somehow these two plugins cannot work together.

**Comment by akildemir** [comment]
My app was not crashing actually, it just wasn't showing the fingerprint dialog, and I realized that it is because of isAvaible() method was not returning the string "ok" in my phone or in whole android 8 platform. It was returning the string "finger". :D. yea just finger. I fucking loled when I first saw it.

**Comment by NiklasMerz** [comment]
Ok good to know

**Comment by fideliocc** [comment]
I'm currently having the same issue. I tested in a blank project and on my early dev project and all worked good. Now is stopping the app when faio.show() runs. Even I can see in console .isAvailable() results but no errors, then app crashes. You have any idea of this behavior? Might I think is dependency conflict maybe? @NiklasMerz 

I attach the logcat output below:
```
01-10 14:44:40.667 25268 25268 E AndroidRuntime: Process: com.ionicframework.conferenceapp, PID: 25268

01-10 14:44:40.673  3170  4880 W ActivityManager: crash : com.ionicframework.conferenceapp,0

01-10 14:44:40.675  3170  4880 W ActivityManager:   Force finishing activity com.ionicframework.conferenceapp/.MainActivity

01-10 14:44:40.677  3170  4880 W MultiScreenManagerService: moveTaskBackToDisplayIfNeeded(): root activity or app is null, task=TaskRecord{6f78821d0 #1318 A=com.ionicframework.conferenceapp U=0 StackId=1 sz=1}, rootActivity=null

01-10 14:44:40.690  3170  3205 I ActivityManager: Showing crash dialog for package com.ionicframework.conferenceapp u0

01-10 14:44:40.756  3170  3205 V WindowManager: Relayout Window{f48b6af u0 Application Error: com.ionicframework.conferenceapp}: viewVisibility=0 req=1014x393 WM.LayoutParams{(0,0)(wrapxwrap) gr=#11 sim=#120 ty=2003 fl=#1820002 pfl=0x1000110 fmt=-3 wanim=0x10302ec vsysui=0x10 surfaceInsets=Rect(6, 6 - 6, 6) needsMenuKey=2 colorMode=0 naviIconColor=0}

01-10 14:44:40.759   713   713 I SurfaceFlinger: id=411 createSurf (1026x405),1 flag=4, Application Error: com.ionicframework.conferenceapp#0

01-10 14:44:40.806  3170  4899 D PackageManager: getComponentMetadataForIconTray : com.ionicframework.conferenceapp.MainActivity does not exist in mServices

01-10 14:44:40.806  3170  4899 D PackageManager: getComponentMetadataForIconTray : com.ionicframework.conferenceapp.MainActivity does not exist in mProviders

01-10 14:44:40.806  3170  4899 D PackageManager: getComponentMetadataForIconTray : com.ionicframework.conferenceapp.MainActivity does not exist in mReceivers

01-10 14:44:40.807  3610 10415 I ApplicationPackageManager: load=com.ionicframework.conferenceapp, bg=144-144, dr=126-126, forDefault=true, density=0

01-10 14:44:40.808  3610 10415 I ApplicationPackageManager: load=com.ionicframework.conferenceapp-crop, bg=126-126, dr=126-126

01-10 14:44:40.818  3170  3205 V WindowManager: finishDrawingLocked: mDrawState=COMMIT_DRAW_PENDING Window{f48b6af u0 Application Error: com.ionicframework.conferenceapp} in Surface(name=Application Error: com.ionicframework.conferenceapp)

01-10 14:44:40.868   713   713 I SurfaceFlinger:      Device | 0x7a859df080 | 0002 | RGBA_8888 |   0.0    0.0 1080.0 2220.0 |    0    0 1080 2220 | com.ionicframework.conferenceapp/com[...]amework.conferenceapp.MainActivity#2

01-10 14:44:40.868   713   713 I SurfaceFlinger:      Device | 0x7a859df480 | 0000 | RGBA_8888 |   0.0    0.0 1026.0  405.0 |   68  892 1011 1264 | Application Error: com.ionicframework.conferenceapp#0

01-10 14:44:40.884   713   713 I SurfaceFlinger:      Device | 0x7a859df080 | 0002 | RGBA_8888 |   0.0    0.0 1080.0 2220.0 |    0    0 1080 2220 | com.ionicframework.conferenceapp/com[...]amework.conferenceapp.MainActivity#2

01-10 14:44:40.884   713   713 I SurfaceFlinger:      Device | 0x7a859df480 | 0000 | RGBA_8888 |   0.0    0.0 1026.0  405.0 |   58  888 1022 1269 | Application Error: com.ionicframework.conferenceapp#0

01-10 14:44:40.890 25734 26166 W PkgUtils: p: com.ionicframework.conferenceapp, u:0

01-10 14:44:41.180  3170  3204 W ActivityManager: Activity pause timeout for ActivityRecord{80bf51f u0 com.ionicframework.conferenceapp/.MainActivity t1318 f}

01-10 14:44:41.254   713   713 I SurfaceFlinger:      Device | 0x7a859df080 | 0002 | RGBA_8888 |   0.0    0.0 1080.0 2220.0 |    2    4 1078 2216 | com.ionicframework.conferenceapp/com[...]amework.conferenceapp.MainActivity#2

01-10 14:44:41.254   713   713 I SurfaceFlinger:      Device | 0x7a859df480 | 0000 | RGBA_8888 |   0.0    0.0 1026.0  405.0 |   27  876 1053 1281 | Application Error: com.ionicframework.conferenceapp#0

01-10 14:44:41.265   713   713 I SurfaceFlinger:      Device | 0x7a859df080 | 0002 | RGBA_8888 |   0.0    0.0 1080.0 2220.0 |    8   16 1072 2203 | com.ionicframework.conferenceapp/com[...]amework.conferenceapp.MainActivity#2

01-10 14:44:41.265   713   713 I SurfaceFlinger: | RGBA_8888 |   0.0    0.0 1026.0  405.0 |   27  876 1053 1281 | Application Error: com.ionicframework.conferenceapp#0

01-10 14:44:41.563   713   713 I SurfaceFlinger:      Device | 0x7a859df480 | 0000 | RGBA_8888 |   0.0    0.0 1026.0  405.0 |   27  876 1053 1281 | Application Error: com.ionicframework.conferenceapp#0

01-10 14:44:51.617  3170  3204 W ActivityManager: Activity destroy timeout for ActivityRecord{80bf51f u0 com.ionicframework.conferenceapp/.MainActivity t1318 f}

01-10 14:44:51.620  3170  3204 I WindowManager: Destroying surface Surface(name=com.ionicframework.conferenceapp/com.ionicframework.conferenceapp.MainActivity) called by com.android.server.wm.WindowStateAnimator.destroySurface:2465 com.android.server.wm.WindowStateAnimator.destroySurfaceLocked:967 com.android.server.wm.WindowState.removeImmediately:2396 com.android.server.wm.WindowState.removeIfPossible:2585 com.android.server.wm.WindowState.removeIfPossible:2425 com.android.server.wm.WindowToken.removeAllWindowsIfPossible:143 com.android.server.wm.AppWindowToken.removeIfPossible:606 com.android.server.wm.AppWindowToken.onRemovedFromDisplay:674 

01-10 14:44:51.623   713  1335 I SurfaceFlinger: id=409 Removed com.ionicframework.conferenceapp/com.ionicframework.conferenceapp.MainActivity#2 (3/9)

01-10 14:44:51.627   713  1335 I SurfaceFlinger: id=409 Removed com.ionicframework.conferenceapp/com.ionicframework.conferenceapp.MainActivity#2 (-2/9)

01-10 14:44:51.639   713   713 I Layer   : id=409 onRemoved com.ionicframework.conferenceapp/com.ionicframework.conferenceapp.MainActivity#2 
```

thanks for your support.

**Comment by NiklasMerz** [comment]
Do you have any other plugins that can cause this? I cannot see anything in the logcat that explains this.

**Comment by fideliocc** [comment]
This is my package.json file:
```
{
  "name": "ionic-conference-app",
  "version": "0.0.0",
  "description": "Ionic Conference App",
  "license": "Apache-2.0",
  "scripts": {
    "ng": "ng",
    "start": "ng serve",
    "build": "ng build",
    "test": "ng test",
    "lint": "ng lint",
    "precommit": "npm run lint"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/ionic-team/ionic-conference-app.git"
  },
  "private": true,
  "dependencies": {
    "@angular/common": "^7.1.4",
    "@angular/core": "^7.1.4",
    "@angular/forms": "^7.1.4",
    "@angular/http": "^7.1.4",
    "@angular/platform-browser": "^7.1.4",
    "@angular/platform-browser-dynamic": "^7.1.4",
    "@angular/pwa": "0.11.0",
    "@angular/router": "~7.1.0",
    "@angular/service-worker": "~7.1.0",
    "@ionic-native/core": "^5.0.0-beta.22",
    "@ionic-native/fingerprint-aio": "^5.0.0-beta.23",
    "@ionic-native/geolocation": "^5.0.0-beta.22",
    "@ionic-native/google-maps": "^5.0.0-beta.25",
    "@ionic-native/in-app-browser": "^5.0.0-beta.23",
    "@ionic-native/pin-dialog": "^5.0.0-beta.22",
    "@ionic-native/splash-screen": "5.0.0-beta.15",
    "@ionic-native/status-bar": "^5.0.0-beta.23",
    "@ionic/angular": "4.0.0-beta.16",
    "@ionic/storage": "^2.1.3",
    "cordova-android": "7.1.4",
    "cordova-browser": "5.0.4",
    "cordova-ios": "^4.5.5",
    "cordova-plugin-device": "^2.0.2",
    "cordova-plugin-fingerprint-aio": "1.5.0",
    "cordova-plugin-geolocation": "4.0.1",
    "cordova-plugin-googlemaps": "git+https://github.com/mapsplugin/cordova-plugin-googlemaps.git#multiple_maps",
    "cordova-plugin-inappbrowser": "^3.0.0",
    "cordova-plugin-ionic-keyboard": "^2.1.3",
    "cordova-plugin-ionic-webview": "^2.3.1",
    "cordova-plugin-pin-dialog": "0.1.3",
    "cordova-plugin-splashscreen": "^5.0.2",
    "cordova-plugin-statusbar": "^2.4.2",
    "cordova-plugin-whitelist": "^1.3.3",
    "core-js": "^2.6.1",
    "rxjs": "6.3.3",
    "sw-toolbox": "3.6.0",
    "zone.js": "^0.8.26"
  },
  "devDependencies": {
    "@angular-devkit/architect": "0.11.0",
    "@angular-devkit/build-angular": "^0.11.4",
    "@angular-devkit/core": "7.1.0",
    "@angular-devkit/schematics": "7.1.0",
    "@angular/cli": "^7.1.4",
    "@angular/compiler": "^7.1.4",
    "@angular/compiler-cli": "^7.1.4",
    "@angular/language-service": "^7.1.4",
    "@ionic/angular-toolkit": "1.2.0",
    "@ionic/lab": "^1.0.16",
    "@types/jasmine": "2.8.11",
    "@types/jasminewd2": "2.0.6",
    "@types/node": "10.12.9",
    "codelyzer": "^4.4.4",
    "husky": "^0.14.3",
    "jasmine-core": "2.99.1",
    "jasmine-spec-reporter": "4.2.1",
    "karma": "3.1.1",
    "karma-chrome-launcher": "2.2.0",
    "karma-coverage-istanbul-reporter": "2.0.4",
    "karma-jasmine": "1.1.2",
    "karma-jasmine-html-reporter": "0.2.2",
    "protractor": "5.4.1",
    "ts-node": "7.0.1",
    "tslint": "^5.12.0",
    "typescript": "~3.1.6"
  },
  "cordova": {
    "platforms": [
      "android",
      "ios",
      "browser"
    ],
    "plugins": {
      "cordova-plugin-device": {},
      "cordova-plugin-inappbrowser": {},
      "cordova-plugin-splashscreen": {},
      "cordova-plugin-whitelist": {},
      "cordova-plugin-ionic-webview": {
        "ANDROID_SUPPORT_ANNOTATIONS_VERSION": "27.+"
      },
      "cordova-plugin-statusbar": {},
      "cordova-plugin-ionic-keyboard": {},
      "cordova-plugin-pin-dialog": {},
      "cordova-plugin-fingerprint-aio": {}
    }
  }
}
```

**Comment by NiklasMerz** [comment]
Maybe the cordova-plugin-pin-dialog plugin is interfering.

**Comment by fideliocc** [comment]
still not working. Might I consider another dependency? Is a really strange behavior. I'll appreciate your support, is important for this project.

**Comment by NiklasMerz** [comment]
Did you test the plugin in a new blank app? Are you running on a real device or emulator?

**Comment by fideliocc** [comment]
I tested it in new blank project and even in my early dev project, and all was working good at the beginning. All tests have been performed on Android devices (Samsung S9+, S8, S7).

**Comment by NiklasMerz** [comment]
Are you really shure the fingeprint plugin causes this crash? Does clearing the app storage on Android help?

**Comment by fideliocc** [comment]
I'm not pretty sure if is the fingerprint plugin itself or is a compatibility issue with another plugin, but app is crashing when fingerprint aio .show() method comes around. I've been checking dependencies, config.xml, testing again and cleared the app cache and storage, but still the same behavior. This is how the class is implemented:

```
export class SendMoneyPage {
  constructor(
    public modalCtrl: ModalController,
    public router: Router,
    private platform: Platform,
    private fingerprint: FingerprintAIO
    ) { }

  async setFingerprint(data) {
    /*** FINGERPRINT AIO ***/
  try {
      await this.platform.ready();
      const available = await this.fingerprint.isAvailable();
      console.log(available);
      if (available === 'finger' || available === 'face') {
        await this.fingerprint.show({
          clientId: 'ionic-conference-app',
          clientSecret: 'password',
          disableBackup: false
        })
        .then((result: any) => {
          console.log(result);
          this.openModal();
        })
        .catch((error: any) => {
          console.log(error);
        });
      }
  } catch (e) {
    console.error(e);
  }
}

  async openModal() {
    const modal = await this.modalCtrl.create({
      component: SendMoneyDialogPage
    });
    await modal.present();
  }
}
```
here are some screenshots from the app and the console output:
![whatsapp image 2019-01-22 at 14 27 09](https://user-images.githubusercontent.com/26985626/51560475-9b1c8580-1e52-11e9-88e6-1a45756401b5.jpeg)

![captura de pantalla 2019-01-22 a la s 14 24 07](https://user-images.githubusercontent.com/26985626/51560496-a8397480-1e52-11e9-8a4c-e5865f6d1e0c.png)

**Comment by NiklasMerz** [comment]
Just a guess but you are doing an unecessary `await`

 ```typescript
**await** this.fingerprint.show({
          clientId: 'ionic-conference-app',
          clientSecret: 'password',
          disableBackup: false
        })
        .then((result: any) => {
          console.log(result);
          this.openModal();
        })
        .catch((error: any) => {
          console.log(error);
        });
````
If you are using then on a promise the await could cause problems. 

You could also try the newest PR #142 to test if that helps for you. Now the app should not crash with certain UI operations.

**Comment by fideliocc** [comment]
I removed and upgraded the cordova-plugin-fingerprint-aio version to 1.6.0 and it works! Thanks for your support.

**Comment by NiklasMerz** [comment]
Perfect. 1.6.0 should become a new release soon

**Comment by JrSagas** [comment]
> I solved this problem by removing this plugin https://ionicframework.com/docs/native/keychain-touch-id/. Somehow these two plugins cannot work together.

It's right!!
Both can't be together, if you try it show this message:

/platforms/android/app/src/main/res/drawable/ic_fingerprint_error.xml" already exists!

Someone, can work with both?
I use keychaintouchid for storage the password.
It works perfectly in iOS, buth not in Android.

Any suggest?

**Comment by NiklasMerz** [comment]
@JrSagas Thank you for reporting. I just created a seperate issue for tracking this.

It looks like those two plugins can work together if we rename some icons. Would you like to try this for solving you problem? If it works please create a pull request. Renaming those files with `git mv` and changing the names in `plugin.xml` should be enough.

Looking forward to your reply.

**Comment by leitonet** [comment]
I'm using ionic + cordova. I solved it adding this in the config.xml:

```
<config-file parent="application" target="AndroidManifest.xml">
            <activity android:exported="true" android:name="de.niklasmerz.cordova.biometric.BiometricActivity" android:theme="@style/TransparentTheme" />
</config-file>
```

**Comment by ShashankHegde16** [comment]
@NiklasMerz  I have installed latest Cordova-android 9.0.0 and Cordova finger print aio3.0.1, its working perfectly on IOS device, but android app is getting crashing as soon as biometric.show method gets triggered. please let me know if there is any work around.
![WhatsApp Image 2020-10-18 at 1 22 26 AM](https://user-images.githubusercontent.com/47588129/96352416-972d9e00-10e0-11eb-8fc6-c37efbc43984.jpeg)

**Comment by ShashankHegde16** [comment]
> I'm using ionic + cordova. I solved it adding this in the config.xml:
> 
> ```
> <config-file parent="application" target="AndroidManifest.xml">
>             <activity android:exported="true" android:name="de.niklasmerz.cordova.biometric.BiometricActivity" android:theme="@style/TransparentTheme" />
> </config-file>
> ```

Not working

**Comment by zimejin** [comment]
> @NiklasMerz I have installed latest Cordova-android 9.0.0 and Cordova finger print aio3.0.1, its working perfectly on IOS device, but android app is getting crashing as soon as biometric.show method gets triggered. please let me know if there is any work around.
> ![WhatsApp Image 2020-10-18 at 1 22 26 AM](https://user-images.githubusercontent.com/47588129/96352416-972d9e00-10e0-11eb-8fc6-c37efbc43984.jpeg)

Facing the same issue as well, please did you find a solution?

**Comment by ShashankHegde16** [comment]
Yes. Upgrade to Latest fingerprint aio version

**Comment by NiklasMerz** [comment]
Please check Logcat in Android Studio if you see any errors and stacktraces

**Reply by zimejin** [reply]
Thanks, I'll do that, I've already updated it to the latest version as well. A quick note is that it works perfectly on any Android Device below Version 10. Also works on IOS last I checked. But crashes on Android 10+
